Update molecule viewer and SMARTs pattern matching

ChiCheng45 opened the pull request:
**Description of work**

- The molecule viewer closes when the button is pressed more than once, in line with other helper dialogs. 
- Bond types are determined using the rdDetermineBonds.DetermineBondOrders in rdkit.
- Molecule viewer now shows the correct bond types.
- SMARTs pattern matching is now working without the need for ~ bond wildcards. SMARTS pattern matching ignore any bonds to dummy atoms.

**Molecule viewer**
Before
<img width="408" height="414" alt="image" src="https://github.com/user-attachments/assets/1a66696d-207b-43c3-8e9b-9ce2c43381d1" />

After
<img width="495" height="419" alt="image" src="https://github.com/user-attachments/assets/1e68b7b0-635b-418f-9411-0178ddf749c3" />

Unfortunately rdDetermineBonds.DetermineBondOrders has some issues dealing with large molecules. Currently, we skip large molecules to avoid long load times. I think there could be a way to reduce the computation time for larger molecules. I have reached out to the RDKit teams to see what they think [#7922](https://github.com/rdkit/rdkit/issues/7922). 

A few things I've noticed which I have left unchanged from the protos (screenshot are from protos).

- Sometimes atoms may be grouped into a molecule but are not bonded.
<img width="438" height="430" alt="Screenshot 2025-10-22 at 09 56 04" src="https://github.com/user-attachments/assets/0d7e499b-0e3f-45bf-87be-0a031dbc99d6" />

- Bonds to dummy atoms will be drawn.
<img width="456" height="430" alt="Screenshot 2025-10-22 at 09 56 56" src="https://github.com/user-attachments/assets/5dec3d3a-565f-4e94-ac23-a2479c518442" />




**Fixes**
Closes #286
Fixes #1030
SMARTS pattern should now work with small/medium molecules.
Molecule viewer should display the correct bond types for small/medium molecules. 

**To test**
Check that the molecule viewer no longer duplicates the window when clicked multiple times. Try out different pattern matches in the atom selection.

Take a look at https://www.daylight.com/dayhtml_tutorials/languages/smarts/smarts_examples.html and let me know if there are any functional groups that would be worth adding.

ChiCheng45 committed:
fixed molecule preview button close, preview diagrams now have bond types and added bond types to rdkit_mol so smarts pattern match now works.

ChiCheng45 committed:
fixed molecule preview button close, preview diagrams now have bond types and added bond types to rdkit_mol so smarts pattern match now works.

ChiCheng45 committed:
improve performance with large molecules. Handle dummy atoms better.

ChiCheng45 committed:
refactored

ChiCheng45 committed:
added lots of checking to ensure that the bond type determination skips molecules with large valence combinations.

ChiCheng45 committed:
fixed docstring

ChiCheng45 committed:
updated combinations calculation

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
fix tests

ChiCheng45 committed:
fix ruff

ChiCheng45 committed:
applied ruff

ChiCheng45 committed:
test fixes

ChiCheng45 committed:
applied ruff

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -93,6 +104,11 @@ def initialise_atoms(
+        self._rdkit_dummy_atms = []
+        for atom in self.rdkit_mol.GetAtoms():
+            if atom.GetAtomicNum() == 0:
+                self._rdkit_dummy_atms.append(atom.GetIdx())
```
```suggestion
        self._rdkit_dummy_atms = [
            atom.GetIdx()
            for atom in self.rdkit_mol.GetAtoms()
            if atom.GetAtomicNum() == 0
        ]
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -102,9 +118,153 @@ def add_atom(self, atm_num: int) -> int:
+        # has two possible valences 3 or 4 when determining the bond types,
+        # DetermineBondOrders will check all possible combinations of valences
+        # see https://github.com/rdkit/rdkit/blob/master/Code/GraphMol/DetermineBonds/DetermineBonds.cpp
+        atomic_valences = {
```
Does this want to go somewhere atoms database-y?

ChiCheng45 replied:
Not right now, this was mostly added to determine whether there would be a lot of valence combinations and to avoid calling the computationally expensive rdkit code. I think there is a fix to their algorithm https://github.com/rdkit/rdkit/issues/7922, hopefully it could be implemented and this bit could be removed.

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -102,9 +118,153 @@ def add_atom(self, atm_num: int) -> int:
+        # deal with them.
+        for cluster_name in self._clusters:
+            for idx, cluster in enumerate(self._clusters[cluster_name]):
+                cluster_no_dummes = [
```
```suggestion
                cluster_no_dummies = [
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -102,9 +118,153 @@ def add_atom(self, atm_num: int) -> int:
+                    element_list[i]
+                    for i, j in enumerate(atm_nums)
+                    if j not in atomic_valences
+                ]
```
```suggestion
                unsupported = [
                    element_list[idx]
                    for idx, atom in enumerate(atm_nums)
                    if atom not in atomic_valences
                ]
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -102,9 +118,153 @@ def add_atom(self, atm_num: int) -> int:
+                    num_bonds = self.rdkit_mol.GetAtomWithIdx(j).GetDegree()
+                    atm_num_valences.append(
+                        len([k for k in atomic_valences[atm_nums[i]] if num_bonds <= k])
+                    )
```
```suggestion
                for atm_num, j in zip(atom_nums, cluster_no_dummes, strict=False):
                    num_bonds = self.rdkit_mol.GetAtomWithIdx(j).GetDegree()
                    atm_num_valences.append(
                        more_itertools.quantify(
                            atomic_valences[atom_num],
                            pred=lambda k: num_bonds <= k
                        )
                    )
```

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/MoleculePreviewWidget.py:
```diff
@@ -94,14 +94,16 @@ def prepare_rdkit_molecule(self) -> chem.RWMol | None:
+            if rdkit_bond is None:
+                submolecule.AddBond(i, j, chem.rdchem.BondType.UNSPECIFIED)
+            else:
+                submolecule.AddBond(i, j, rdkit_bond.GetBondType())
```
```suggestion
            bond_type = rdkit_bond.GetBondType() if rdkit_bond is not None else chem.rdchem.BondType.UNSPECIFIED
            submolecule.AddBond(i, j, bond_type)
```

oerc0122 reviewed (approved):
Looks good, just a couple of potential stylistic things.

ChiCheng45 committed:
Update MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Chi Cheng <chi-yang.cheng@proton.me>

ChiCheng45 committed:
Update MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Chi Cheng <chi-yang.cheng@proton.me>

ChiCheng45 committed:
Update MDANSE_GUI/Src/MDANSE_GUI/InputWidgets/MoleculePreviewWidget.py

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Chi Cheng <chi-yang.cheng@proton.me>

ChiCheng45 committed:
fix spelling

MBartkowiakSTFC commented:
I tried several trajectories, but for every single one so far I have been getting:
```
2025-10-22 12:16:05,814 - INFO - process[21847] - ChemicalSystem 246 - Determining bond orders for molecule H2_O1 with index 0
2025-10-22 12:16:05,817 - ERROR - process[21847] - ChemicalSystem 256 - Error determining bond orders for molecule H2_O1 with index 0: Python argument types in
    rdkit.Chem.rdDetermineBonds.DetermineBondOrders(RWMol)
did not match C++ signature:
    DetermineBondOrders(RDKit::ROMol {lvalue} mol, int charge=0, bool allowChargedFragments=True, bool embedChiral=True, bool useAtomMap=False). SMARTS pattern matching will not work as expected. Bond types set to UNSPECIFIED use bond type wildcard ~ to match bonds in this molecule.
```

ChiCheng45 committed:
apply suggestion

ChiCheng45 commented:
> I tried several trajectories, but for every single one so far I have been getting:
> 
> ```
> 2025-10-22 12:16:05,814 - INFO - process[21847] - ChemicalSystem 246 - Determining bond orders for molecule H2_O1 with index 0
> 2025-10-22 12:16:05,817 - ERROR - process[21847] - ChemicalSystem 256 - Error determining bond orders for molecule H2_O1 with index 0: Python argument types in
>     rdkit.Chem.rdDetermineBonds.DetermineBondOrders(RWMol)
> did not match C++ signature:
>     DetermineBondOrders(RDKit::ROMol {lvalue} mol, int charge=0, bool allowChargedFragments=True, bool embedChiral=True, bool useAtomMap=False). SMARTS pattern matching will not work as expected. Bond types set to UNSPECIFIED use bond type wildcard ~ to match bonds in this molecule.
> ```

What version of rdkit are you using? Here is the one I am using.

rdkit                            2025.3.5         pypi_0                   pypi

ChiCheng45 committed:
fix ruff

ChiCheng45 committed:
fix ruff

ChiCheng45 committed:
apply ruff

MBartkowiakSTFC commented:
> 
> What version of remit are you using? Here is the one I am using.
> 
> rdkit 2025.3.5 pypi_0 pypi

Mine was 2024.9.5 and I thought it would have got updated when I did `pip install --upgrade` on MDANSE itself, but in the end I had to upgrade it manually. Now I have rdkit 2025.9.1 and the bond detection works.

MBartkowiakSTFC reviewed (commented):
Looks good, definitely works better than before.

If you wanted to add some other pre-defined SMARTS strings to the current set, my preference would be:
1. Carboxylic acid
`[CX3](=O)[OX2H1]`
2. Unfused benzene ring
`[cR1]1[cR1][cR1][cR1][cR1][cR1]1`

Should we have a lower limit on the version of `rdkit` in our `pyproject.toml`? You could set it to the version you are using now, since we know that it works, and so does the latest one.

Also, interesting to see if the RdKit authors accept your suggested changes.

oerc0122 commented:
> Looks good, definitely works better than before.
> 
> If you wanted to add some other pre-defined SMARTS strings to the current set, my preference would be:
> 
> 1. Carboxylic acid
>    `[CX3](=O)[OX2H1]`
> 2. Unfused benzene ring
>    `[cR1]1[cR1][cR1][cR1][cR1][cR1]1`
> 
> Should we have a lower limit on the version of `rdkit` in our `pyproject.toml`? You could set it to the version you are using now, since we know that it works, and so does the latest one.
> 
> Also, interesting to see if the RdKit authors accept your suggested changes.

Do we at some point want some sort of setting to allow users to define their own custom groups? Definitely not this PR, but might be worth an issue.

ChiCheng45 committed:
set rdkit version in pyproject

MBartkowiakSTFC commented:
> 
> Do we at some point want some sort of setting to allow users to define their own custom groups? Definitely not this PR, but might be worth an issue.

I added this as issue #1040

MBartkowiakSTFC commented:
Will this close #286?

ChiCheng45 commented:
> Will this close #286?

Yes this will close #286.

ChiCheng45 commented:
> Looks good, definitely works better than before.
> 
> If you wanted to add some other pre-defined SMARTS strings to the current set, my preference would be:
> 
> 1. Carboxylic acid
>    `[CX3](=O)[OX2H1]`
> 2. Unfused benzene ring
>    `[cR1]1[cR1][cR1][cR1][cR1][cR1]1`
> 
> Should we have a lower limit on the version of `rdkit` in our `pyproject.toml`? You could set it to the version you are using now, since we know that it works, and so does the latest one.
> 
> Also, interesting to see if the RdKit authors accept your suggested changes.

I've added both groups above. I've changed them slightly so that the carboxylic acid group also matches the hydrogen.

<img width="1445" height="392" alt="image" src="https://github.com/user-attachments/assets/a74ccd93-34e1-41aa-b2d7-c93fe777b5bf" />

And the unfused benzene matches all hydrogens are bonded to it. The SMARTS pattern is "\[cR1r6\]([H]).[cR1r6]".

<img width="276" height="221" alt="image" src="https://github.com/user-attachments/assets/22f214c0-3dea-4615-867c-6e8a451b8dc8" />

ChiCheng45 committed:
adjusted set_bond_orders and fixed for ring system smarts pattern matching. Add unfused benzene and carboxylic acid group to predefined smarts patterns.

ChiCheng45 committed:
update smart pattern qline edit default to first selection

ChiCheng45 committed:
ruff fix

MBartkowiakSTFC reviewed (commented):
At the moment I have been testing it on the trajectories from SlimMD, and they all have large molecules for which rdkit fails. Then I can still create a selection using SMARTS, but I need to insert `~` every time in the strings. Do you think we need to add some extra hints for the users who are not familiar with SMARTS, to show them where these should be added? You already mention it in the log message. We could add an example to the current list of pre-defined SMARTS strings, or add some note in the documentation, if you think it is meaningful.

ChiCheng45 committed:
update documentation and warning log

ChiCheng45 committed:
update documentation and warning log

ChiCheng45 commented:
> At the moment I have been testing it on the trajectories from SlimMD, and they all have large molecules for which rdkit fails. Then I can still create a selection using SMARTS, but I need to insert `~` every time in the strings. Do you think we need to add some extra hints for the users who are not familiar with SMARTS, to show them where these should be added? You already mention it in the log message. We could add an example to the current list of pre-defined SMARTS strings, or add some note in the documentation, if you think it is meaningful.

I've updated to warning log and updated to documentation. Also, I think we could increase the max number of atoms that the bond type determination but I wanted to make sure that it never causes any performance issues. Perhaps, we should add some sort of setting to allow the user to increase that max number of atoms for bond determination, or move it so that it is done in the trajectory editor or something else.

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -102,9 +119,122 @@ def add_atom(self, atm_num: int) -> int:
+                # dummy atoms.
+                self.rdkit_mol.AddBond(i, j, Chem.rdchem.BondType.UNSPECIFIED)
+
+    def set_bond_orders(self, coords: np.ndarray, max_iters=1000, max_natms=100):
```
```suggestion
    def set_bond_orders(self, coords: np.ndarray, *, max_iters: int = 1000, max_natms: int = 100):
```

oerc0122 commented on MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py:
```diff
@@ -93,6 +104,12 @@ def initialise_atoms(
         if name_list is not None:
             self._atom_names = [str(x) for x in name_list]
 
+        self._rdkit_dummy_atms = [
```
If we're only ever checking `__contains__` (`in`) do we want this to be a set rather than a list (constant time vs. linear)

oerc0122 commented on MDANSE_GUI/Src/MDANSE_GUI/Widgets/SelectionWidgets.py:
```diff
@@ -496,7 +499,7 @@ def add_specific_widgets(self):
         self.selection_field.addItems(self.pattern_dictionary.keys())
         layout.addWidget(QLabel("pattern:"))
-        self.input_field = QLineEdit("", self)
+        self.input_field = QLineEdit(next(iter(self.pattern_dictionary.values())), self)
```
```suggestion
        self.input_field = QLineEdit(more_itertools.first(self.pattern_dictionary.values()), self)
```

ChiCheng45 committed:
Update MDANSE/Src/MDANSE/Chemistry/ChemicalSystem.py

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Chi Cheng <chi-yang.cheng@proton.me>

ChiCheng45 committed:
Update MDANSE_GUI/Src/MDANSE_GUI/Widgets/SelectionWidgets.py

Co-authored-by: Jacob Wilkins <46597752+oerc0122@users.noreply.github.com>
Signed-off-by: Chi Cheng <chi-yang.cheng@proton.me>

ChiCheng45 committed:
apply suggestions and fix ruff

MBartkowiakSTFC reviewed (approved):
Thanks for the changes, looks good.
